packages/x-data-grid/src/hooks/features/rows/useGridRowSpanning.ts

```diff
@@ -322,35 +310,41 @@ export const useGridRowSpanning = (
         };
       });
     },
-    [apiRef, props.rowSpanning, range, renderContext, visibleRows, colDefs, processedRange],
+    [apiRef, processedRange, props.pagination, props.paginationMode],
   );
 
-  const prevRenderContext = React.useRef(renderContext);
-  const isFirstRender = React.useRef(true);
-  const shouldResetState = React.useRef(false);
-  const previousTree = React.useRef(tree);
-  React.useEffect(() => {
-    const firstRender = isFirstRender.current;
-    if (isFirstRender.current) {
-      isFirstRender.current = false;
-    }
-    if (tree !== previousTree.current) {
-      previousTree.current = tree;
-      updateRowSpanningState(true);
+  // Reset events trigger a full re-computation of the row spanning state:
+  // - The `unstable_rowSpanning` prop is updated (feature flag)
+  // - The filtering is applied
+  // - The sorting is applied
+  // - The `paginationModel` is updated
+  // - The rows are updated
+  const onRowsUpdate = React.useCallback(() => {
```

Comment:
It doesn't run on row update only, maybe a name closer to it's behavior?
```suggestion
  const recomputeRowSpanningState = React.useCallback(() => {
```

Reply:
Agreed. Maybe `resetRowSpanningState` instead to align it with the `resetState` param in the handler it calls.